README.md
NuLib2
======

NuLib2 is a command-line file archiver for Apple II archives.  It can
operate on ShrinkIt and Binary II files (.shk, .sdk, .bxy, .bse, .bny, .bqy).

NufxLib is a library of code that supports ShrinkIt archives.  It's
used by NuLib2 and CiderPress.

A pre-built NuLib2 binary is available for x86 Windows.  For Linux and
Mac OS X systems, you can download the source code and build it.

More information, including full documentation for NuLib2 and NufxLib,
can be found on http://www.nulib.com/.



nulib2/ChangeLog.txt
2017/09/21  ***** v3.1.0 shipped *****

2015/12/26  fadden
    - Fix handling of entries with missing threads.
    - Improve handling of Mac OS X file type attributes.
    - Updated "recognized extensions" table.

2015/01/09  ***** v3.0.0 shipped *****

2015/01/03  fadden
    - Mac OS X: get file/aux type from FinderInfo when adding files.
    - Mac OS X: set file type and creator when extracting from Binary ][.

2015/01/02  fadden
    - Distinguish Unicode and Mac OS Roman strings.

2014/12/22  fadden
    - Source code cleanup.

2014/10/30  ***** v2.2.2 shipped *****

2014/10/28  fadden
    - Switched from CVS on sourceforge to github.
    - Updated configure scripts and makefiles.

2009/01/13  fadden
    - Handle files wrapped in Binary ][ by ProTERM.
    - (This was marked as 2.2.1, but not delivered as a formal release.)

2007/02/19  ***** v2.2.0 shipped *****

2007/02/19  fadden
    - Enable "bad Mac" handling.
    - Switched from GPL to BSD license.

2006/02/18  ***** v2.1.1 shipped *****

2006/02/18  fadden
    - Fix handling of MS-DOS reserved names.  Besides handling names like
      "con", we also need to handle "con.foo.txt".

2005/09/17  ***** v2.1.0 shipped *****

2004/10/11  ***** v2.0.3 shipped *****

2004/03/10  ***** v2.0.2 shipped *****

2003/10/16  ***** v2.0.1 shipped *****

2003/03/18  ***** v2.0.0 shipped *****

2003/02/18  fadden
    - When extracting with "-ee", disk images now have ".PO" appended.
    - Resurrected HandleAddNotFound().
    - Switched to case-sensitive filename comparisons.

2003/02/08  fadden
    - Upped version to v2.0.0.
    - Many fixes to pathname handling:
        - Correctly handle '%' when preservation is OFF.
        - Accept 4-char extensions in '-ee' without risk of buffer overflow.
        - Fixed broken assert when converting long %xx names.
        - Store "AUX" as "%00AUX" when preserving Win32 names (vs. "_AUX").
        - Always store files with ':' as path separator.
        - Recognize that some Win32 variants (Win2K and later at the least)
          will accept both '/' and '\' as pathname separators.
        - Correctly convert ".//foo" to "foo" instead of "/foo".
    - Tracked changes to NufxLib DataSource API.

2003/01/10  fadden
    - Check NufxLib "compiled" version against "linked" version.

2002/12/06  fadden
    - Made minor changes to allow linking NufxLib in as a DLL.

2002/10/20  ***** v1.1.0 shipped *****

2002/10/10  fadden
    - added fancy help text (-h)

2002/10/09  fadden
    - added "-zz" flag to specify libbz2's "bzip2" compression

2002/10/08  fadden
    - added Binary II support

2002/09/30  fadden
    - added "-z" flag to specify zlib's "deflate" compression (the
      "secret" debug dump command is now -g)

2002/09/26  fadden
    - progress updater now shows "analyzing" for scan pass of SQ

2002/09/23  fadden
    - ran the code through valgrind; found and fixed some minor bugs

2002/09/20  fadden
    - pulled the sources out and started fiddling with them again
    - changed hard tabs to spaces

2000/05/22  ***** v1.0.1 shipped (no changes - version follows nufxlib) *****

2000/05/18  ***** v1.0.0 shipped *****

2000/05/18  fadden
    - added nulib2 to set of things stripped by "distbin"
    - updated version information to indicate final release

2000/03/25  ***** v0.6.1 shipped *****

2000/03/25      fadden
    - Sheppy says Mac OS X PPC v1.02 and v1.2 work with minor SysDefs tweak

2000/03/05  ***** v0.6.0 (beta) shipped *****

2000/03/05  fadden
    - don't call mktemp(), just pass template into NuOpenRW
    - removed DEBUG_MSGS from default CFLAGS
    - updated version information to indicate beta release

2000/02/24  ***** v0.5.1 shipped *****

2000/02/20  changes from Scott Blackman
    - portability fixes for DJGPP under Win95

2000/02/17  changes from Devin Reade
    - portability fixes for BSD, AIX, and others
    - added "distbin" target

2000/02/09  ***** v0.5.0 (alpha) shipped *****

2000/02/09  fadden
    - changed the comparison used when extracting/deleting a list of files
      from strcasecmp to strcmp, since NufxLib does case-sensitive compares.
    - fixed the percentage for files and archives larger than 21MB

2000/02/08  fadden
    - tweaked the BeOS/PPC config around a little
    - deleted some commas to make "gcc -pedantic" happy
    - changed version to x.y.z format here too
    - generalized the "aux" handling to include all MS-DOS device names

2000/02/06  fadden
    - include @CFLAGS@ in case somebody wants to override them

2000/02/06  ***** v0.4b shipped *****

2000/02/06  fadden
    - added "install-shared" make target
    - portability fixes for HP/UX

2000/02/06  ***** v0.4a shipped *****

2000/02/06  fadden
    - massaged configure.in for BeOS, and added some type casts for mwerks

2000/02/06  ***** v0.4 shipped *****

2000/02/05  fadden
    - added "mkinstalldirs" to install target
    - added Win32 makefile
    - made a few implicit typecasts explicit for Visual C++'s benefit
    - change "aux" to "_aux", because FAT filesystems choke on it

2000/02/04  fadden
    - added Win32 recursive directory descent

2000/02/02  fadden
    - minor changes to get it working under Win32 (Visual C++ 6.0)
    - added --enable-dmalloc to configuration

2000/02/01  fadden
    - screen out leading "./", and junk the path if ".." shows up in path
    - don't try to add comments to records we're skipping
    - set kNuValueEOL appropriately for the current system

2000/01/29  ***** v0.3 shipped *****

2000/01/29  fadden
    - added "make install" target, with the standard autoconf defines
    - added some examples to the man page

2000/01/28  fadden
    - merged "Kind" and "Type" columns in "v" output
    - display a '+' when doing EOL conversions on an extracted file

2000/01/26  fadden
    - added UI for allowing the user to ignore bad CRCs
    - implemented "-j" (junk paths) for add and extract
    - implemented "-c" (comments) for add and extract

nufxlib/Debug.c
/*
 * NuFX archive manipulation library
 * Copyright (C) 2000-2007 by Andy McFadden, All Rights Reserved.
 * This is free software; you can redistribute it and/or modify it under the
 * terms of the BSD License, see the file COPYING-LIB.
 *
 * Debugging functions.  These are omitted from the non-debug build.
 */
#include "NufxLibPriv.h"

#if defined(DEBUG_MSGS)


/* pull a string out of one of the static arrays */
#define GetStaticString(index, staticArray) ( \
            (index) >= NELEM(staticArray) ? "<unknown>" : staticArray[index] \
        )

/* thread's thread_class */
static const char* gThreadClassNames[] = {
    "message_thread",
    "control_thread",
    "data_thread",
    "filename_thread",
};

/* thread's thread_format */
static const char* gThreadFormatNames[] = {
    "uncompressed",
    "Huffman Squeeze",
    "dynamic LZW/1",
    "dynamic LZW/2",
    "12-bit LZC",
    "16-bit LZC",
    "deflate",
    "bzip2"
};

/* days of the week */
static const char* gDayNames[] = {
    "[ null ]",
    "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
};
/* months of the year */
static const char* gMonths[] = {
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
};

#define kNuDateOutputLen    64

/* file_sys_id values */
static const char* gFileSysIDs[] = {
    "Reserved/unknown ($00)", "ProDOS/SOS", "DOS 3.3", "DOS 3.2",
    "Apple II Pascal", "Macintosh (HFS)", "Macintosh (MFS)",
    "LISA file system", "Apple CP/M", "Reserved 0x09", "MS-DOS",
    "High-Sierra", "ISO 9660", "AppleShare"
};


/*
 * Convert a DateTime structure into something printable.
 *
 * The buffer passed in must hold at least kNuDateOutputLen bytes.
 *
 * Returns "buffer" for the benefit of printf() calls.
 */
static char* Nu_DebugDumpDate(const NuDateTime* pDateTime, char* buffer)
{
    char* cp;

    /* is it valid? */
    if (pDateTime->day > 30 || pDateTime->month > 11 || pDateTime->hour > 24 ||
        pDateTime->minute > 59)
    {
        strcpy(buffer, "   <invalid>   ");
        goto bail;
    }

    /* is it empty? */
    if ((pDateTime->second | pDateTime->minute | pDateTime->hour |
        pDateTime->year | pDateTime->day | pDateTime->month |
        pDateTime->extra | pDateTime->weekDay) == 0)
    {
        strcpy(buffer, "   [No Date]   ");
        goto bail;
    }

    cp = buffer;

    /* only print weekDay if one was stored */
    if (pDateTime->weekDay) {
        if (pDateTime->weekDay < NELEM(gDayNames))
            sprintf(cp, "%s, ", gDayNames[pDateTime->weekDay]);
        else
            sprintf(cp, "??%d, ", pDateTime->weekDay);
        cp += strlen(cp);
    }

    sprintf(cp, "%02d-%s-%04d  %02d:%02d:%02d",
        pDateTime->day+1, gMonths[pDateTime->month],
        pDateTime->year < 40 ? pDateTime->year + 2000 : pDateTime->year + 1900,
        pDateTime->hour, pDateTime->minute, pDateTime->second);

bail:
    sprintf(buffer + strlen(buffer), "  [s%d m%d h%d Y%d D%d M%d x%d w%d]",
        pDateTime->second, pDateTime->minute, pDateTime->hour,
        pDateTime->year, pDateTime->day, pDateTime->month, pDateTime->extra,
        pDateTime->weekDay);

    return buffer;
}


/*
 * Convert a buffer into a hexadecimal character string.
 *
 * The result will be 2x the size of the original, +1 for a null byte.
 */
static void ConvertToHexStr(const uint8_t* inBuf, int inLen, char* outBuf)
{
    while (inLen--) {
        *outBuf++ = HexConv((*inBuf >> 4) & 0x0f);
        *outBuf++ = HexConv(*inBuf & 0x0f);
        inBuf++;
    }
    *outBuf = '\0';
}


/*
 * Dump everything we know about pThread.
 */
void Nu_DebugDumpThread(const NuThread* pThread)
{
    static const char* kInd = "      ";
    NuThreadID threadID;
    const char* descr;

    Assert(pThread != NULL);

    printf("%sThreadClass:  0x%04x (%s)\n", kInd,
        pThread->thThreadClass,
        GetStaticString(pThread->thThreadClass, gThreadClassNames));
    printf("%sThreadFormat: 0x%04x (%s)\n", kInd,
        pThread->thThreadFormat,
        GetStaticString(pThread->thThreadFormat, gThreadFormatNames));

    threadID = NuMakeThreadID(pThread->thThreadClass, pThread->thThreadKind);
    switch (threadID) {
    case kNuThreadIDOldComment: descr = "old comment";  break;
    case kNuThreadIDComment:    descr = "comment";      break;
    case kNuThreadIDIcon:       descr = "icon";         break;
    case kNuThreadIDMkdir:      descr = "mkdir";        break;
    case kNuThreadIDDataFork:   descr = "data fork";    break;
    case kNuThreadIDDiskImage:  descr = "disk image";   break;
    case kNuThreadIDRsrcFork:   descr = "rsrc fork";    break;
    case kNuThreadIDFilename:   descr = "filename";     break;
    default:                    descr = "<unknown>";    break;
    }
    printf("%sThreadKind:   0x%04x (%s)\n", kInd,
        pThread->thThreadKind, descr);

    printf("%sThreadCRC: 0x%04x  ThreadEOF: %lu  CompThreadEOF: %lu\n", kInd,
        pThread->thThreadCRC, pThread->thThreadEOF, pThread->thCompThreadEOF);
    printf("%s*File data offset: %ld  actualThreadEOF: %ld\n", kInd,
        pThread->fileOffset, pThread->actualThreadEOF);
}

/*
 * Dump everything we know about pRecord, including its threads and ThreadMods.
 *
 * Changes to existing records are made to the "copy" set, not the "orig"
 * set.  Pass in the "orig" copy in "pRecord", and optionally pass in the
 * "copy" set in "pXrefRecord" to glean data from both.
 */
static void Nu_DebugDumpRecord(NuArchive* pArchive, const NuRecord* pRecord,
    const NuRecord* pXrefRecord, Boolean isDeleted)
{
    NuError err;    /* dummy */
    static const char* kInd = "    ";
    char dateBuf[kNuDateOutputLen];
    const NuThreadMod* pThreadMod;
    const NuThread* pThread;
    uint32_t idx;

    Assert(pRecord != NULL);

    /*printf("PTR: pRecord=0x%08lx pXrefRecord=0x%08lx\n", (long) pRecord,
        (long) pXrefRecord);*/

    printf("%s%s%sFilename: '%s' (idx=%lu)\n", kInd,
        isDeleted ? "[DEL] " : "",
        pXrefRecord != NULL && pXrefRecord->pThreadMods != NULL ? "[MOD] " : "",
        pRecord->filename == NULL ? "<not specified>" : pRecord->filename,
        pRecord->recordIdx);
    printf("%sHeaderID: '%.4s'  VersionNumber: 0x%04x  HeaderCRC: 0x%04x\n",
        kInd,
        pRecord->recNufxID, pRecord->recVersionNumber, pRecord->recHeaderCRC);
    printf("%sAttribCount: %u  TotalThreads: %lu\n", kInd,
        pRecord->recAttribCount, pRecord->recTotalThreads);
    printf("%sFileSysID: %u (%s)  FileSysInfo: 0x%04x ('%c')\n", kInd,
        pRecord->recFileSysID,
        GetStaticString(pRecord->recFileSysID, gFileSysIDs),
        pRecord->recFileSysInfo,
        NuGetSepFromSysInfo(pRecord->recFileSysInfo));
    /* do something fancy for ProDOS? */
    printf("%sFileType: 0x%08lx  ExtraType: 0x%08lx  Access: 0x%08lx\n", kInd,
        pRecord->recFileType, pRecord->recExtraType, pRecord->recAccess);
    printf("%sCreateWhen:  %s\n", kInd,
        Nu_DebugDumpDate(&pRecord->recCreateWhen, dateBuf));
    printf("%sModWhen:     %s\n", kInd,
        Nu_DebugDumpDate(&pRecord->recModWhen, dateBuf));
    printf("%sArchiveWhen: %s\n", kInd,
        Nu_DebugDumpDate(&pRecord->recArchiveWhen, dateBuf));
    printf("%sStorageType: %u  OptionSize: %u  FilenameLength: %u\n", kInd,
        pRecord->recStorageType, pRecord->recOptionSize,
        pRecord->recFilenameLength);
    if (pRecord->recOptionSize) {
        char* outBuf = Nu_Malloc(pArchive, pRecord->recOptionSize * 2 +1);
        BailAlloc(outBuf);
        Assert(pRecord->recOptionList != NULL);
        ConvertToHexStr(pRecord->recOptionList, pRecord->recOptionSize, outBuf);
        printf("%sOptionList: [%s]\n", kInd, outBuf);
        Nu_Free(pArchive, outBuf);
    }

    printf("%s*ExtraCount: %ld  RecFileOffset: %ld  RecHeaderLength: %ld\n",
        kInd,
        pRecord->extraCount, pRecord->fileOffset, pRecord->recHeaderLength);

    for (idx = 0; idx < pRecord->recTotalThreads; idx++) {
        Boolean isFake;

        isFake = (idx >= pRecord->recTotalThreads - pRecord->fakeThreads);
        pThread = Nu_GetThread(pRecord, idx);
        Assert(pThread != NULL);

        printf("%s--Thread #%lu (idx=%lu)%s\n", kInd, idx, pThread->threadIdx,
            isFake ? " [FAKE]" : "");
        Nu_DebugDumpThread(pThread);
    }

    if (pXrefRecord != NULL)
        pThreadMod = pXrefRecord->pThreadMods;
    else
        pThreadMod = pRecord->pThreadMods;  /* probably empty */

    if (pThreadMod != NULL)
        printf("%s*ThreadMods -----\n", kInd);
    while (pThreadMod != NULL) {
        switch (pThreadMod->entry.kind) {
        case kNuThreadModAdd:
            printf("%s  *-ThreadMod ADD 0x%08lx 0x%04x (sourceType=%d)\n", kInd,
                pThreadMod->entry.add.threadID,
                pThreadMod->entry.add.threadFormat,
                Nu_DataSourceGetType(pThreadMod->entry.add.pDataSource));
            break;
        case kNuThreadModUpdate:
            printf("%s  *-ThreadMod UPDATE %6ld\n", kInd,
                pThreadMod->entry.update.threadIdx);
            break;
        case kNuThreadModDelete:
            printf("%s  *-ThreadMod DELETE %6ld\n", kInd,
                pThreadMod->entry.delete.threadIdx);
            break;
        case kNuThreadModUnknown:
        default:
            Assert(0);
            printf("%s++ThreadMod UNKNOWN\n", kInd);
            break;
        }

        pThreadMod = pThreadMod->pNext;
    }

    /*printf("%s*TotalLength: %ld  TotalCompLength: %ld\n",
        kInd, pRecord->totalLength, pRecord->totalCompLength);*/
    printf("%s*TotalCompLength: %ld\n", kInd, pRecord->totalCompLength);
    printf("\n");

bail:
    return;
}

/*
 * Dump the records in a RecordSet.
 */
static void Nu_DebugDumpRecordSet(NuArchive* pArchive,
    const NuRecordSet* pRecordSet, const NuRecordSet* pXrefSet)
{
    const NuRecord* pRecord;
    const NuRecord* pXrefRecord;
    Boolean doXref;
    long count;

    doXref = false;
    pXrefRecord = NULL;
    if (pXrefSet != NULL && Nu_RecordSet_GetLoaded(pXrefSet)) {
        pXrefRecord = Nu_RecordSet_GetListHead(pXrefSet);
        doXref = true;
    }

    /* dump every record, if we've loaded them */
    count = Nu_RecordSet_GetNumRecords(pRecordSet);
    pRecord = Nu_RecordSet_GetListHead(pRecordSet);
    if (pRecord != NULL) {
        Assert(count != 0);
        while (count--) {
            Assert(pRecord != NULL);

            if (pXrefRecord != NULL &&
                pRecord->recordIdx == pXrefRecord->recordIdx)
            {
                Nu_DebugDumpRecord(pArchive, pRecord, pXrefRecord, false);
                pXrefRecord = pXrefRecord->pNext;
            } else {
                Nu_DebugDumpRecord(pArchive, pRecord, NULL, doXref);
            }
            pRecord = pRecord->pNext;
        }
    } else {
        Assert(count == 0);
    }
}

/*
 * Dump the master header block.
 */
static void Nu_DebugDumpMH(const NuMasterHeader* pMasterHeader)
{
    static const char* kInd = "  ";
    char dateBuf1[kNuDateOutputLen];

    Assert(pMasterHeader != NULL);

    printf("%sNufileID: '%.6s'  MasterCRC: 0x%04x  TotalRecords: %lu\n", kInd,
        pMasterHeader->mhNufileID, pMasterHeader->mhMasterCRC,
        pMasterHeader->mhTotalRecords);
    printf("%sArchiveCreateWhen: %s\n", kInd,
        Nu_DebugDumpDate(&pMasterHeader->mhArchiveCreateWhen, dateBuf1));
    printf("%sArchiveModWhen:    %s\n", kInd,
        Nu_DebugDumpDate(&pMasterHeader->mhArchiveModWhen, dateBuf1));
    printf("%sMasterVersion: %u  MasterEOF: %lu\n", kInd,
        pMasterHeader->mhMasterVersion, pMasterHeader->mhMasterEOF);
}

/*
 * Dump everything we know about pArchive.
 *
 * This will only print the records that we have seen so far.  If the
 * application hasn't caused us to scan through all of the records in
 * the archive, then this won't be very interesting.  This will never
 * show any records for streaming-mode archives.
 */
void Nu_DebugDumpAll(NuArchive* pArchive)
{
    Assert(pArchive != NULL);

    printf("*Archive pathname: '%s'\n", pArchive->archivePathname);
    printf("*Archive type: %d\n", pArchive->archiveType);
    printf("*Header offset: %ld (junk offset=%ld)\n",
        pArchive->headerOffset, pArchive->junkOffset);
    printf("*Num records: %ld orig, %ld copy, %ld new\n",
        Nu_RecordSet_GetNumRecords(&pArchive->origRecordSet),
        Nu_RecordSet_GetNumRecords(&pArchive->copyRecordSet),
        Nu_RecordSet_GetNumRecords(&pArchive->newRecordSet));
    printf("*NuRecordIdx seed: %lu  NuRecordIdx next: %lu\n",
        pArchive->recordIdxSeed, pArchive->nextRecordIdx);

    /* master header */
    Nu_DebugDumpMH(&pArchive->masterHeader);

    printf("  *ORIG record set (x-ref with COPY):\n");
    Nu_DebugDumpRecordSet(pArchive, &pArchive->origRecordSet,
        &pArchive->copyRecordSet);
    printf("  *NEW record set:\n");
    Nu_DebugDumpRecordSet(pArchive, &pArchive->newRecordSet, NULL);

    if (!Nu_RecordSet_GetLoaded(&pArchive->origRecordSet) &&
        !Nu_RecordSet_GetLoaded(&pArchive->newRecordSet))
    {
        printf("*** DEBUG: original records not loaded yet? ***\n");
    }

}

#endif /*DEBUG_MSGS*/
